aggregator/src/aggregator/collection_job_driver.rs

```diff
@@ -302,18 +310,26 @@ where
                         &batch_aggregations,
                     );
 
-                    try_join!(
-                        try_join_all(
-                            batch_aggregations
-                                .iter()
-                                .map(|ba| tx.update_batch_aggregation(ba))
-                        ),
-                        try_join_all(
-                            empty_batch_aggregations
-                                .iter()
-                                .map(|ba| tx.put_batch_aggregation(ba))
-                        ),
-                    )?;
+                    // Rather than awaiting all the futures at once, which can cause heap growth
+                    // proportional to the number of batch aggregations to write, put them into an
+                    // buffered stream so that they get run in groups of bounded size. #3857
+                    futures::stream::iter(
+                        batch_aggregations
+                            .iter()
+                            // existing batch aggregatons must be updated so zip with true
+                            .zip(repeat(true))
+                            // empty batch aggregations are new so zip with false
+                            .chain(empty_batch_aggregations.iter().zip(repeat(false))),
+                    )
+                    .map(Result::Ok)
```

Comment:
Similarly, we could replace this stream combinator with an iterator combinator
```suggestion
                            .chain(empty_batch_aggregations.iter().zip(repeat(false)))
                            .map(Result::Ok),
                    )
```